### PR TITLE
balance account creation changes for new approval type

### PR DESCRIPTION
**Pull request by bflood-strike**
Balance Account Creation Approval Type
## Description

- Updated the existing Balance Account Init code
- Can set policy plus initial settings (whitelist/dapps enabled) during the Init.
- Also made the change to add the new balance accounts source address into the address book.
- the change is not backward compatible since we updated the existing instruction definition

## Motivation and Context

Needed for new approval types epic.

## How Has This Been Tested?

Unit tests
Will also test outside with integration tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Bug fix breaking (fix that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ x] New feature breaking (change which adds functionality and causes existing functionality to change)

## Checklist:

- [ x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ x] I have added tests to cover my changes.
- [ x] All new and existing tests passed.

### CODE REVIEW

**Review by aonyshchenko-strike — COMMENTED** [review]
Looks great! 👍

**Review by aonyshchenko-strike — APPROVED** [review]

File: src/handlers/balance_account_creation_handler.rs [hunk 1]
```diff
@@ -25,7 +25,7 @@ pub fn init(
 
     let wallet = Wallet::unpack(&wallet_account_info.data.borrow())?;
     wallet.validate_config_initiator(initiator_account_info)?;
-    wallet.validate_add_balance_account(account_guid_hash, update)?;
+    wallet.validate_balance_account_creation(account_guid_hash, initial_account, program_id)?;
```
[the remark this answers is not in the document]
Comment (bflood-strike) [reply src/handlers/balance_account_creation_handler.rs]: thanks

File: src/model/wallet.rs [hunk 2]
```diff
@@ -314,32 +323,65 @@ impl Wallet {
...
-        self_clone.add_balance_account(account_guid_hash, update)
+        self_clone.create_balance_account(account_guid_hash, initial_account, program_id)
     }
 
-    pub fn add_balance_account(
+    pub fn create_balance_account(
         &mut self,
         account_guid_hash: &BalanceAccountGuidHash,
-        update: &BalanceAccountUpdate,
+        initial_account: &BalanceAccountCreation,
```
[the remark this answers is not in the document]
Comment (bflood-strike) [reply src/model/wallet.rs]: ok, globally changes initial_account to creation_params

File: src/model/wallet.rs [hunk 3]
```diff
@@ -314,32 +323,65 @@ impl Wallet {
...
+        self.enable_transfer_approvers(balance_account_idx, &initial_account.transfer_approvers)?;
+
+        let (source_account_pda, _) =
+            Pubkey::find_program_address(&[&account_guid_hash.to_bytes()], program_id);
+
+        self.add_address_book_entries(&vec![(
+            initial_account.address_book_slot_id,
+            AddressBookEntry {
+                address: source_account_pda,
+                name_hash: AddressBookEntryNameHash::new(array_ref![
```
[the remark this answers is not in the document]
Comment (bflood-strike) [reply src/model/wallet.rs]: changed this per your suggestion

**Review by aonyshchenko-strike — APPROVED** [review]
Looks great! 👍

**Pull request merged**